### DISCUSSION TITLE
Is it possible to only use filesystem caching for builds that aren't re-builds?

### DISCUSSION THREAD

**Discussion by mkmmnn** [webpack/webpack#18950] [Q&A]
**Is it possible to only use filesystem caching for builds that aren't re-builds?**

I recently upgraded a project to use Webpack 5.95.0 and am running into performance issues with my development builds. Without filesystem caching on, build times are significantly (>50% slower) than on Webpack 4. Turning on filesystem caching alleviates some of this pain by reducing build times when the cache is warm such that the average build time is reasonable.

However, in development (watch mode on), I'm observing higher recompile times when filesystem caching is turned on. I'm suspecting this is because time is spent dealing with the filesystem caching, though I'm not sure.

I'd like to keep benefitting from the filesystem cache on full builds, but I want to avoid writing or reading from the cache in watch mode. Then just write to the cache when shutting down webpack. Is this possible?

**Comment by alexander-akait** [comment]
@mkmmnn We have https://webpack.js.org/configuration/cache/#cachereadonly, and `WEBPACK_SERVE` in `webpack.config.js`, so you can enable/disable it

**Reply by mkmmnn** [reply]
Thanks for getting back to me!

I have looked into the `cache.readonly` configuration option, but I've gleaned both empirically and from reading the docs that this option won't write to the filesystem cache when I shut down webpack. I would only be able to get slightly closer to my desired behavior by following these steps:
- Build with `cache.type = filesystem` and `cache.readonly = false` to warm the filesystem cache
- Quit webpack
- Rebuild with `cache.type = filesystem` and `cache.readonly = true` (development mode, so `watch = true`) so watch-mode recompiles don't need to deal with unpacking + repacking the cache
- If I want to update the filesystem cache, do a new build with `cache.type = filesystem` and `cache.readonly = false`

In theory, this could work to get me both faster average build times due to the warm filesystem cache _and_ faster watch-mode recompile times. However, this isn't a practical solution for my team because it requires each developer to devote additional time to keeping track of which webpack build settings they ought to use. And it requires more builds overall to keep the filesystem cache fresh.

Is it possible to just set watch-mode recompiles to `readonly`? But alongside that, to write to the cache when I close the development webpack process?

**Reply by alexander-akait** [reply]
You can use https://github.com/streamich/memfs in watch mode, we do the same in https://github.com/webpack/webpack-dev-middleware/blob/master/src/utils/setupOutputFileSystem.js#L64, just create a simple plugin that runs at the initialization stage (hook) and setup memfs, memory usage should speed up the process, if you need to write some files to disk - use this https://github.com/webpack/webpack-dev-middleware/blob/master/src/utils/setupWriteToDisk.js#L29
